packages/amplify-category-auth/provider-utils/awscloudformation/triggers/CreateAuthChallenge/function-template-dir/trigger-index.js

```diff
@@ -1,14 +1,58 @@
-/*
-  this file will loop through all js modules which are uploaded to the lambda resource,
-  provided that the file names (without extension) are included in the "MODULES" env variable.
-  "MODULES" is a comma-delimmited string.
-*/
+/**
+ * @fileoverview
+ *
+ * This CloudFormation Trigger creates a Lambda function which waits for all
+ * other specified Lambdas (modules) to resolve, which should be located in the
+ * same directory as this file (./).
+ *
+ * IMPORTANT: This handler *must* return the `event` since it may be consumed by
+ * additional layers in a CloudFormation stack. Failing to return an acceptable
+ * event may cause errors like "Invalid JSON" and prevent properly configured
+ * projects from deploying as expected.
+ *
+ * @see https://github.com/aws-amplify/amplify-cli/pull/7219
+ */
+
+/**
+ * The names of modules to load are stored as a comma-delimited string in the
+ * `MODULES` env variable.
+ */
 const moduleNames = process.env.MODULES.split(',');
+/**
+ * The array of imported modules.
+ */
 const modules = moduleNames.map(name => require(`./${name}`));
 
-exports.handler = (event, context, callback) => {
-  for (let i = 0; i < modules.length; i += 1) {
-    const { handler } = modules[i];
-    handler(event, context, callback);
-  }
+/**
+ * This handler is itself a Lambda function which iterates over all of the given
+ * modules and awaits them.
+ *
+ * @see https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html
+ *
+ * @param {object} event
+ *
+ * The event that triggered this Lambda.
+ *
+ * @param {object} context
+ *
+ * The context for this Lambda. See:
+ * https://docs.aws.amazon.com/lambda/latest/dg/nodejs-context.html
+ *
+ * @param {function} callback
```

Comment:
Since this PR is migrating to `async/await` style, are the leftover callbacks still needed?

Reply:
I left it in for backwards compatibility. In the linked issues, the way we all worked around was with a `context.done()` call, so there is utility in not immediately ripping it out IMO. If someone passes a `callback`, it will work as expected - I see no reason to violate that.

Reply:
> I left it in for backwards compatibility.

My understanding is that this code is only generated once, and that users updating their projects will not regenerate it. If that's the case, then I think we should really discourage using both callbacks and async/await.

Reply:
I don't think it's immediately obvious what we should do here:

- If we remove it, we basically kill support for any old patterns (new projects will be generated with the new template but old patterns exist in guides and blog posts published on AWS), and we'd want to make it very clear to users that the handlers generated by the CLI will use the async/await form, throw any errors, return values directly.
- If we don't remove it, old code using the callback will still work as written, but we might not be able to guarantee execution order (I think?)

Reply:
The more I think about this, the more I think it's a bad idea to intermingle callbacks and promises like this. For example, the trigger-index files will now be expecting the modules it invokes to return a value to indicate completion. However, if one of those modules does something like this, I think it could cause problems:

```js
setTimeout(() => { callback(); }, 1000);
// Implicit return
```

Reply:
Hi @ctjlewis. Just following up - are you planning to remove the callbacks in this PR?